src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -274,15 +623,292 @@ TryDropShard(GroupShardPlacement *placement)
 									   dropQuery->data);
 
 	/* remove the shard from the node */
-	bool success =
-		SendOptionalCommandListToWorkerOutsideTransaction(shardPlacement->nodeName,
-														  shardPlacement->nodePort,
-														  NULL, dropCommandList);
-	if (success)
+	bool success = SendOptionalCommandListToWorkerOutsideTransaction(nodeName,
+																	 nodePort,
+																	 NULL,
+																	 dropCommandList);
+
+	return success;
+}
+
+
+/*
+ * GetNextOperationId allocates and returns a unique operationId for an operation
+ * requiring potential cleanup. This allocation occurs both in shared memory and
+ * in write ahead logs; writing to logs avoids the risk of having operationId collisions.
+ */
+static OperationId
+GetNextOperationId()
+{
+	OperationId operationdId = INVALID_OPERATION_ID;
+
+	/*
+	 * In regression tests, we would like to generate operation IDs consistently
+	 * even if the tests run in parallel. Instead of the sequence, we can use
+	 * the next_operation_id GUC to specify which operation ID the current session should
+	 * generate next. The GUC is automatically increased by 1 every time a new
+	 * operation ID is generated.
+	 */
+	if (NextOperationId > 0)
 	{
-		/* delete the actual placement */
-		DeleteShardPlacementRow(placement->placementId);
+		operationdId = NextOperationId;
+		NextOperationId += 1;
+
+		return operationdId;
 	}
 
-	return success;
+	/* token location, or -1 if unknown */
+	const int location = -1;
+	RangeVar *sequenceName = makeRangeVar(PG_CATALOG,
+										  OPERATIONID_SEQUENCE_NAME,
+										  location);
+
+	bool missingOK = false;
+	Oid sequenceId = RangeVarGetRelid(sequenceName, NoLock, missingOK);
+
+	bool checkPermissions = false;
+	operationdId = nextval_internal(sequenceId, checkPermissions);
```

Comment:
not for this pr, but we should probably change our other nextval calls to this pattern instead of switching to superuser

Reply:
@marcocitus , This needs to be fixed to use a subtransaction as it will hang replication slot creation.